pkg/storegateway/series_chunks.go

```diff
@@ -4,46 +4,148 @@ package storegateway
 
 import (
 	"context"
+	"sync"
 	"time"
 
 	"github.com/pkg/errors"
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/prometheus/model/labels"
 
 	"github.com/grafana/mimir/pkg/storegateway/storepb"
+	util_math "github.com/grafana/mimir/pkg/util/math"
 	"github.com/grafana/mimir/pkg/util/pool"
 )
 
+// Mimir compacts blocks up to 24h. Assuming a 5s scrape interval as worst case scenario,
+// and 120 samples per chunk, there could be 86400 * (1 / 5) * (1 / 120) = 144 chunks for
+// a series in the biggest block. Using a slab size of 1000 looks a good trade-off to support
+// high frequency scraping without wasting too much memory in case of queries hitting a low
+// number of chunks (across series).
+const seriesChunksSlabSize = 1000
+
+var (
+	seriesEntrySlicePool = pool.Interface(&sync.Pool{
+		// Intentionally return nil if the pool is empty, so that the caller can preallocate
+		// the slice with the right size.
+		New: nil,
+	})
+
+	seriesChunksSlicePool = pool.Interface(&sync.Pool{
+		// Intentionally return nil if the pool is empty, so that the caller can preallocate
+		// the slice with the right size.
+		New: nil,
+	})
+)
+
 // seriesChunksSetIterator is the interface implemented by an iterator returning a sequence of seriesChunksSet.
 type seriesChunksSetIterator interface {
 	Next() bool
+
+	// At returns the current seriesChunksSet. The caller should (but NOT must) invoke seriesChunksSet.release()
+	// on the returned set once it's guaranteed it will not be used anymore.
 	At() seriesChunksSet
+
 	Err() error
 }
 
 // seriesChunksSet holds a set of series, each with its own chunks.
 type seriesChunksSet struct {
-	series []seriesEntry
+	series           []seriesEntry
+	seriesReleasable bool
+
+	// It gets lazy initialized (only if required).
+	seriesChunksPool *pool.SlabPool[storepb.AggrChunk]
 
 	// chunksReleaser releases the memory used to allocate series chunks.
 	chunksReleaser chunksReleaser
 }
 
+// newSeriesChunksSet creates a new seriesChunksSet. The series slice is pre-allocated with
+// the provided seriesCapacity at least. This means this function GUARANTEES the series slice
+// will have a capacity of at least seriesCapacity.
+//
+// If seriesReleasable is true, then a subsequent call release() will put the internal
+// series slices to a memory pool for reusing.
+func newSeriesChunksSet(seriesCapacity int, seriesReleasable bool) seriesChunksSet {
+	var prealloc []seriesEntry
+
+	// If it's releasable then we try to reuse a slice from the pool.
+	if seriesReleasable {
+		if reused := seriesEntrySlicePool.Get(); reused != nil {
```

Comment:
a thought more than a suggestion: I liked how in other places we inject the pool. Like the chunk bytes pool for the chunks Raw data. It will make tests less dependant on each other (if you write one bad test you can fail a lot of them), slightly more consistent, and a bit more clear that we're managing our memory.

I don't have experience with memory management, so if it's generally better to make this pooling as invisible as possible, then your approach is better.

Reply:
`seriesChunksSet` uses two pools and passing them around is a bit annoying. I would do it only if there's a real justification. Currently I've been able to test pooling in a clean way anyway.